includes/class-wc-stripe-intent-controller.php

```diff
@@ -102,6 +102,23 @@ public function verify_intent() {
 		try {
 			$gateway->verify_intent_after_checkout( $order );
 
+			if ( isset( $_GET['save_payment_method'] ) && ! empty( $_GET['save_payment_method'] ) ) {
+				$intent = $gateway->get_intent_from_order( $order );
+				if ( isset( $intent->last_payment_error ) ) {
+					// Currently, Stripe saves the payment method even if the authentication fails for 3DS cards.
+					// Although, the card is not stored in DB we need to remove the source from the customer on Stripe
+					// in order to keep the sources in sync with the data in DB.
+					$customer = new WC_Stripe_Customer( wp_get_current_user()->ID );
+					$customer->delete_source( $intent->last_payment_error->source->id );
```

Comment:
I haven't found a way to tell Stripe to not to save the 3DS card if the authentication fails. I've chatted with Stripe support about this and I'm waiting for an official response. For now I'm deleting the source from the customer manually.